Bottom: fcd5c053af80f8c6a741c9b1880859c12a767a29
Top:    ab6e278d6b99d2be46d28dddd9f6980e1ea2e0f7
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2016-01-09 17:30:20 +0000

polygon.c: Add option to include drills holes (from vias / pins) in the board outline


---

diff --git a/src/hid/gtk/gtkhid-gl.c b/src/hid/gtk/gtkhid-gl.c
index 1c8b92b..1fd999f 100644
--- a/src/hid/gtk/gtkhid-gl.c
+++ b/src/hid/gtk/gtkhid-gl.c
@@ -1617,7 +1617,7 @@ ensure_board_outline (void)
     if (PCB->Data->outline != NULL)
       poly_Free (&PCB->Data->outline);
 
-    PCB->Data->outline = board_outline_poly ();
+    PCB->Data->outline = board_outline_poly (false);
     PCB->Data->outline_valid = true;
   }
 }
diff --git a/src/polygon.c b/src/polygon.c
index 922cb83..ac9f638 100644
--- a/src/polygon.c
+++ b/src/polygon.c
@@ -2023,6 +2023,22 @@ line_outline_callback (const BoxType * b, void *cl)
   return 1;
 }
 
+static int
+pv_outline_callback (const BoxType * b, void *cl)
+{
+  PinType *pv = (PinType *)b;
+  struct clip_outline_info *info = cl;
+  POLYAREA *np, *res;
+
+  if (!(np = CirclePoly (pv->X, pv->Y, pv->DrillingHole / 2)))
+    return 0;
+
+  poly_Boolean_free (info->poly, np, &res, PBO_SUB);
+  info->poly = res;
+
+  return 1;
+}
+
 static void
 delete_piece_cb (gpointer data, gpointer userdata)
 {
@@ -2044,7 +2060,7 @@ delete_piece_cb (gpointer data, gpointer userdata)
   poly_Free (&piece);
 }
 
-POLYAREA *board_outline_poly (void)
+POLYAREA *board_outline_poly (bool include_holes)
 {
   int i;
   int count;
@@ -2114,6 +2130,12 @@ POLYAREA *board_outline_poly (void)
   r_search (Layer->line_tree, &region, NULL, line_outline_callback, &info);
   r_search (Layer->arc_tree,  &region, NULL, arc_outline_callback, &info);
 
+  if (include_holes)
+    {
+      r_search (PCB->Data->pin_tree, &region, NULL, pv_outline_callback, &info);
+      r_search (PCB->Data->via_tree, &region, NULL, pv_outline_callback, &info);
+    }
+
   clipped = info.poly;
 
   /* Now we just need to work out which pieces of polygon are inside
diff --git a/src/polygon.h b/src/polygon.h
index 30e5ea3..a02fd99 100644
--- a/src/polygon.h
+++ b/src/polygon.h
@@ -88,5 +88,5 @@ bool MorphPolygon (LayerType *, PolygonType *);
 void NoHolesPolygonDicer (PolygonType *p, const BoxType *clip,
                           void (*emit) (PLINE *, void *), void *user_data);
 void PolyToPolygonsOnLayer (DataType *, LayerType *, POLYAREA *, FlagType);
-POLYAREA *board_outline_poly (void);
+POLYAREA *board_outline_poly (bool include_holes);
 #endif
